File: client/crypto/message_crypto.py
"""
Message encryption and decryption for E2EE messaging.

Implements forward-secret message encryption using ephemeral X25519 keys,
ECDH key agreement, HKDF key derivation, and AES-256-GCM authenticated encryption.
"""

import json
import os
from base64 import b64decode, b64encode

from cryptography.exceptions import InvalidSignature, InvalidTag
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)
from cryptography.hazmat.primitives.asymmetric.x25519 import (
    X25519PrivateKey,
    X25519PublicKey,
)
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.hkdf import HKDF

from crypto.key_generator import KeyGenerator


class MessageEncryptor:
    """Encrypts messages with forward secrecy using ephemeral keys."""

    def encrypt_message(
        self,
        plaintext: str,
        recipient_x25519_pub: X25519PublicKey,
        sender_ed25519_priv: Ed25519PrivateKey,
        sender_x25519_pub: X25519PublicKey,
        sender_id: str,
        recipient_id: str,
    ) -> dict:
        """
        Encrypt a message for both recipient and sender (double encryption).

        The message key is generated once and encrypted twice:
        - once with a wrapping key derived from ECDH(ephemeral, recipient_pub)  → blob for recipient
        - once with a wrapping key derived from ECDH(ephemeral, sender_pub)     → blob for sender

        Both blobs share the same ciphertext and signature, so E2EE is preserved.

        Returns:
            dict with "blob", "sender_blob", "signature" (all base64 encoded)
        """
        # 1. Generate ephemeral X25519 keypair
        ephemeral_priv, ephemeral_pub = KeyGenerator.generate_ephemeral_keypair()
        ephemeral_pub_bytes = KeyGenerator.serialize_public_key(ephemeral_pub)

        # 2. Encrypt plaintext with a random message key (shared between both copies)
        message_key = os.urandom(32)
        nonce_msg = os.urandom(12)
        aesgcm_msg = AESGCM(message_key)
        ciphertext_msg = aesgcm_msg.encrypt(nonce_msg, plaintext.encode("utf-8"), None)

        # 3. Build recipient blob — message key wrapped with ECDH(ephemeral, recipient_pub)
        recipient_pub_bytes = KeyGenerator.serialize_public_key(recipient_x25519_pub)
        shared_secret_r = ephemeral_priv.exchange(recipient_x25519_pub)
        wrapping_key_r = self._derive_wrapping_key(shared_secret_r, ephemeral_pub_bytes, recipient_pub_bytes)
        nonce_wrap_r = os.urandom(12)
        encrypted_msg_key_r = AESGCM(wrapping_key_r).encrypt(nonce_wrap_r, message_key, None)

        blob_dict = {
            "sender_id": sender_id,
            "recipient_id": recipient_id,
            "ephemeral_pub": b64encode(ephemeral_pub_bytes).decode("ascii"),
            "encrypted_msg_key": b64encode(encrypted_msg_key_r).decode("ascii"),
            "nonce_wrap": b64encode(nonce_wrap_r).decode("ascii"),
            "ciphertext_msg": b64encode(ciphertext_msg).decode("ascii"),
            "nonce_msg": b64encode(nonce_msg).decode("ascii"),
        }
        json_bytes = json.dumps(blob_dict, sort_keys=True, separators=(",", ":")).encode("utf-8")

        # 4. Sign the recipient blob
        signature_bytes = sender_ed25519_priv.sign(json_bytes)

        # 5. Build sender blob — same ciphertext, message key wrapped with ECDH(ephemeral, sender_pub)
        sender_pub_bytes = KeyGenerator.serialize_public_key(sender_x25519_pub)
        shared_secret_s = ephemeral_priv.exchange(sender_x25519_pub)
        wrapping_key_s = self._derive_wrapping_key(shared_secret_s, ephemeral_pub_bytes, sender_pub_bytes)
        nonce_wrap_s = os.urandom(12)
        encrypted_msg_key_s = AESGCM(wrapping_key_s).encrypt(nonce_wrap_s, message_key, None)

        sender_blob_dict = {
            "sender_id": sender_id,
            "recipient_id": recipient_id,
            "ephemeral_pub": b64encode(ephemeral_pub_bytes).decode("ascii"),
            "encrypted_msg_key": b64encode(encrypted_msg_key_s).decode("ascii"),
            "nonce_wrap": b64encode(nonce_wrap_s).decode("ascii"),
            "ciphertext_msg": b64encode(ciphertext_msg).decode("ascii"),
            "nonce_msg": b64encode(nonce_msg).decode("ascii"),
        }
        sender_json_bytes = json.dumps(sender_blob_dict, sort_keys=True, separators=(",", ":")).encode("utf-8")

        return {
            "blob": b64encode(json_bytes).decode("ascii"),
            "sender_blob": b64encode(sender_json_bytes).decode("ascii"),
            "signature": b64encode(signature_bytes).decode("ascii"),
        }

    @staticmethod
    def _derive_wrapping_key(shared_secret: bytes, ephemeral_pub_bytes: bytes, peer_pub_bytes: bytes) -> bytes:
        salt = ephemeral_pub_bytes + peer_pub_bytes
        hkdf = HKDF(algorithm=hashes.SHA256(), length=32, salt=salt, info=b"msg-v1")
        return hkdf.derive(shared_secret)


class MessageDecryptor:
    """Decrypts and verifies encrypted messages."""

    def decrypt_message(
        self,
        encrypted_msg: dict,
        recipient_x25519_priv: X25519PrivateKey,
        sender_ed25519_pub: Ed25519PublicKey,
    ) -> str:
        """Decrypt and verify a message received from another user."""
        blob_bytes = b64decode(encrypted_msg["blob"])
        signature_bytes = b64decode(encrypted_msg["signature"])

        try:
            sender_ed25519_pub.verify(signature_bytes, blob_bytes)
        except InvalidSignature:
            raise InvalidSignature("Message signature verification failed")

        return self._decrypt_blob(blob_bytes, recipient_x25519_priv)

    def decrypt_own_message(
        self,
        sender_blob_b64: str,
        sender_x25519_priv: X25519PrivateKey,
    ) -> str:
        """Decrypt the sender's own copy of a sent message (no signature check needed)."""
        return self._decrypt_blob(b64decode(sender_blob_b64), sender_x25519_priv)

    @staticmethod
    def _decrypt_blob(blob_bytes: bytes, x25519_priv: X25519PrivateKey) -> str:
        blob_dict = json.loads(blob_bytes.decode("utf-8"))

        ephemeral_pub_bytes = b64decode(blob_dict["ephemeral_pub"])
        encrypted_msg_key = b64decode(blob_dict["encrypted_msg_key"])
        nonce_wrap = b64decode(blob_dict["nonce_wrap"])
        ciphertext_msg = b64decode(blob_dict["ciphertext_msg"])
        nonce_msg = b64decode(blob_dict["nonce_msg"])

        ephemeral_pub = KeyGenerator.load_x25519_public_key(ephemeral_pub_bytes)
        shared_secret = x25519_priv.exchange(ephemeral_pub)

        peer_pub_bytes = KeyGenerator.serialize_public_key(x25519_priv.public_key())
        salt = ephemeral_pub_bytes + peer_pub_bytes
        hkdf = HKDF(algorithm=hashes.SHA256(), length=32, salt=salt, info=b"msg-v1")
        wrapping_key = hkdf.derive(shared_secret)

        try:
            message_key = AESGCM(wrapping_key).decrypt(nonce_wrap, encrypted_msg_key, None)
        except InvalidTag:
            raise InvalidTag("Failed to decrypt message key")

        try:
            plaintext_bytes = AESGCM(message_key).decrypt(nonce_msg, ciphertext_msg, None)
        except InvalidTag:
            raise InvalidTag("Failed to decrypt message plaintext")

        return plaintext_bytes.decode("utf-8")

File: flutter_client/lib/crypto/message_crypto.dart
/// Message encryption/decryption using ephemeral X25519 + AES-256-GCM.
///
/// This is the version-1 ("double wrapped") scheme used by the Python client's
/// `MessageEncryptor`/`MessageDecryptor`. The v2 Double-Ratchet path lives in
/// `ratchet_facade.dart`; it reuses [deriveWrappingKey] to build the
/// sender's own stateless copy of each message.
library;

import 'dart:convert';
import 'dart:typed_data';

import 'keys.dart';

/// Derive the AES key that wraps the per-message key for one peer.
Future<Uint8List> deriveWrappingKey(
  List<int> sharedSecret,
  List<int> ephemeralPubBytes,
  List<int> peerPubBytes,
) =>
    hkdf(sharedSecret, <int>[...ephemeralPubBytes, ...peerPubBytes],
        utf8.encode('msg-v1'), keyLen);

class MessageEncryptor {
  /// Encrypt [plaintext] for both recipient and sender, returning base64
  /// `blob`, `sender_blob`, and `signature` (mirrors the Python v1 scheme).
  static Future<Map<String, String>> encryptMessage({
    required String plaintext,
    required List<int> recipientX25519Pub,
    required List<int> senderEd25519Priv,
    required List<int> senderEd25519Pub,
    required List<int> senderX25519Priv,
    required List<int> senderX25519Pub,
    required String senderId,
    required String recipientId,
  }) async {
    final ephemeral = await generateX25519Keypair();
    final ephemeralPubBytes = ephemeral.$2;

    final messageKey = randomBytes(keyLen);
    final nonceMsg = randomBytes(nonceSize);
    final ctMsg = await aesEncrypt(utf8.encode(plaintext), messageKey,
        nonce: nonceMsg);

    // Recipient blob
    final sharedR = await xShared(ephemeral.$1, recipientX25519Pub);
    final wrapR = await deriveWrappingKey(sharedR, ephemeralPubBytes, recipientX25519Pub);
    final nonceWrapR = randomBytes(nonceSize);
    final encMsgKeyR = await aesEncrypt(messageKey, wrapR, nonce: nonceWrapR);

    final blobDict = {
      'sender_id': senderId,
      'recipient_id': recipientId,
      'ephemeral_pub': base64Encode(ephemeralPubBytes),
      'encrypted_msg_key': base64Encode(encMsgKeyR),
      'nonce_wrap': base64Encode(nonceWrapR),
      'ciphertext_msg': base64Encode(ctMsg),
      'nonce_msg': base64Encode(nonceMsg),
    };
    final blobBytes = utf8.encode(canonicalJson(blobDict));
    final signature = await edSign(blobBytes, senderEd25519Priv);

    // Sender blob (same ciphertext, wrapped under sender's long-term key)
    final sharedS = await xShared(ephemeral.$1, senderX25519Pub);
    final wrapS = await deriveWrappingKey(sharedS, ephemeralPubBytes, senderX25519Pub);
    final nonceWrapS = randomBytes(nonceSize);
    final encMsgKeyS = await aesEncrypt(messageKey, wrapS, nonce: nonceWrapS);

    final senderBlobDict = {
      'sender_id': senderId,
      'recipient_id': recipientId,
      'ephemeral_pub': base64Encode(ephemeralPubBytes),
      'encrypted_msg_key': base64Encode(encMsgKeyS),
      'nonce_wrap': base64Encode(nonceWrapS),
      'ciphertext_msg': base64Encode(ctMsg),
      'nonce_msg': base64Encode(nonceMsg),
    };
    final senderBlobBytes = utf8.encode(canonicalJson(senderBlobDict));

    return {
      'blob': base64Encode(blobBytes),
      'sender_blob': base64Encode(senderBlobBytes),
      'signature': base64Encode(signature),
    };
  }
}

class MessageDecryptor {
  static Future<String> decryptMessage(
    Map<String, dynamic> encryptedMsg,
    List<int> recipientX25519Priv,
    List<int> recipientX25519Pub,
    List<int> senderEd25519Pub,
  ) async {
    final blobBytes = base64Decode(encryptedMsg['blob'] as String);
    final signature = base64Decode(encryptedMsg['signature'] as String);
    if (!await edVerify(blobBytes, signature, senderEd25519Pub)) {
      throw Exception('Message signature verification failed');
    }
    return _decryptBlob(blobBytes, recipientX25519Priv, recipientX25519Pub);
  }

  static Future<String> decryptOwnMessage(
    String senderBlobB64,
    List<int> senderX25519Priv,
    List<int> senderX25519Pub,
  ) async {
    return _decryptBlob(base64Decode(senderBlobB64), senderX25519Priv, senderX25519Pub);
  }

  static Future<String> _decryptBlob(
      Uint8List blobBytes, List<int> x25519Priv, List<int> x25519Pub) async {
    final blob = jsonDecode(utf8.decode(blobBytes)) as Map<String, dynamic>;
    final ephemeralPub = base64Decode(blob['ephemeral_pub'] as String);
    final encMsgKey = base64Decode(blob['encrypted_msg_key'] as String);
    final nonceWrap = base64Decode(blob['nonce_wrap'] as String);
    final ctMsg = base64Decode(blob['ciphertext_msg'] as String);
    final nonceMsg = base64Decode(blob['nonce_msg'] as String);

    final shared = await xShared(x25519Priv, ephemeralPub);
    final wrap = await deriveWrappingKey(shared, ephemeralPub, x25519Pub);
    final messageKey = await aesDecrypt(encMsgKey, wrap, nonce: nonceWrap);
    final plaintext = await aesDecrypt(ctMsg, messageKey, nonce: nonceMsg);
    return utf8.decode(plaintext);
  }
}
